File: app/src/main/java/com/mikelle/qufoi/app/model/Theme.java
package com.mikelle.qufoi.app.model;

import android.support.annotation.ColorRes;
import android.support.annotation.StyleRes;

import com.mikelle.qufoi.app.R;

/**
 * A way to make simple changes to the application's appearance at runtime in correlation to its
 * {@link Category}.
 *
 * Usually this should be done via attributes and {@link android.view.ContextThemeWrapper}s.
 * In one case in Qufoi it is more performant to work like this.
 * This case involves a trade-off between statically loading these themes versus inflation
 * in an adapter backed view without recycling.
 */
public enum Theme {
    qufoi(R.color.qufoi_primary, R.color.qufoi_primary_dark,
            R.color.theme_blue_background, R.color.theme_blue_text,
            R.color.qufoi_accent, R.style.Qufoi),
    blue(R.color.theme_blue_primary, R.color.theme_blue_primary_dark,
            R.color.theme_blue_background, R.color.theme_blue_text,
            R.color.theme_blue_accent, R.style.Qufoi_Blue),
    green(R.color.theme_green_primary, R.color.theme_green_primary_dark,
            R.color.theme_green_background, R.color.theme_green_text,
            R.color.theme_green_accent, R.style.Qufoi_Green),
    purple(R.color.theme_purple_primary, R.color.theme_purple_primary_dark,
            R.color.theme_purple_background, R.color.theme_purple_text,
            R.color.theme_purple_accent, R.style.Qufoi_Purple),
    red(R.color.theme_red_primary, R.color.theme_red_primary_dark,
            R.color.theme_red_background, R.color.theme_red_text,
            R.color.theme_red_accent, R.style.Qufoi_Red),
    yellow(R.color.theme_yellow_primary, R.color.theme_yellow_primary_dark,
            R.color.theme_yellow_background, R.color.theme_yellow_text,
            R.color.theme_yellow_accent, R.style.Qufoi_Yellow);

    private final int mColorPrimaryId;
    private final int mWindowBackgroundColorId;
    private final int mColorPrimaryDarkId;
    private final int mTextColorPrimaryId;
    private final int mAccentColorId;
    private final int mStyleId;

    Theme(final int colorPrimaryId, final int colorPrimaryDarkId,
            final int windowBackgroundColorId, final int textColorPrimaryId,
            final int accentColorId, final int styleId) {
        mColorPrimaryId = colorPrimaryId;
        mWindowBackgroundColorId = windowBackgroundColorId;
        mColorPrimaryDarkId = colorPrimaryDarkId;
        mTextColorPrimaryId = textColorPrimaryId;
        mAccentColorId = accentColorId;
        mStyleId = styleId;
    }

    @ColorRes
    public int getTextPrimaryColor() {
        return mTextColorPrimaryId;
    }

    @ColorRes
    public int getWindowBackgroundColor() {
        return mWindowBackgroundColorId;
    }

    @ColorRes
    public int getPrimaryColor() {
        return mColorPrimaryId;
    }

    @ColorRes
    public int getAccentColor() {
        return mAccentColorId;
    }

    @ColorRes
    public int getPrimaryDarkColor() {
        return mColorPrimaryDarkId;
    }

    @StyleRes
    public int getStyleId() {
        return mStyleId;
    }
}

File: app/src/main/java/com/mikelle/qufoi/app/model/Theme.kt
package com.mikelle.qufoi.app.model

import android.support.annotation.ColorRes
import android.support.annotation.StyleRes

import com.mikelle.qufoi.app.R

/**
 * A way to make simple changes to the application's appearance at runtime in correlation to its
 * [Category].

 * Usually this should be done via attributes and [android.view.ContextThemeWrapper]s.
 * In one case in Qufoi it is more performant to work like this.
 * This case involves a trade-off between statically loading these themes versus inflation
 * in an adapter backed view without recycling.
 */
enum class Theme private constructor(val primaryColor: Int, val primaryDarkColor: Int,
                                     val windowBackgroundColor: Int, val textPrimaryColor: Int,
                                     val accentColor: Int, val styleId: Int) {
    qufoi(R.color.qufoi_primary, R.color.qufoi_primary_dark, R.color.theme_blue_background, R.color.theme_blue_text, R.color.qufoi_accent, R.style.Qufoi),
    blue(R.color.theme_blue_primary, R.color.theme_blue_primary_dark, R.color.theme_blue_background, R.color.theme_blue_text, R.color.theme_blue_accent, R.style.Qufoi_Blue),
    green(R.color.theme_green_primary, R.color.theme_green_primary_dark, R.color.theme_green_background, R.color.theme_green_text, R.color.theme_green_accent, R.style.Qufoi_Green),
    purple(R.color.theme_purple_primary, R.color.theme_purple_primary_dark, R.color.theme_purple_background, R.color.theme_purple_text, R.color.theme_purple_accent, R.style.Qufoi_Purple),
    red(R.color.theme_red_primary, R.color.theme_red_primary_dark, R.color.theme_red_background, R.color.theme_red_text, R.color.theme_red_accent, R.style.Qufoi_Red),
    yellow(R.color.theme_yellow_primary, R.color.theme_yellow_primary_dark, R.color.theme_yellow_background, R.color.theme_yellow_text, R.color.theme_yellow_accent, R.style.Qufoi_Yellow)
}
